Update Redis app directory

razvand opened the pull request:
Rework the contents of the Redis app directory to make it more suitable for users (prime tool is KraftKit) and for developers (prime tools are Make, Kconfig, QEMU / Firecracker).

Move and output files in dedicated directories to keep the root directory clean: `kraft.yaml`, `README.md`, `Makefile`, `Makefile.uk` are the only files in the dedicated directories.

Update `README.md` according to the new structure.

Improve support for KraftKit: documentation, run scriipts, testing.

razvand committed:
Use workdir/ as clone directory instead of .unikraft/

`.unikraft/` was previously used as the directory to clone library
repositories (including the `unikraft` core repository). Being a hidden,
it is not immediately available to the developer.

Use `workdir/` instead.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
Use workdir/build/ as output directory

For consistency on placing external and generated files, place the build
directory in `workdir/build/` instead of `build/`.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
README: Use make prepare before build command

The build system may bump into issues when using `make` right out the
configuration.

Use `make prepare` before `make` for a "safe" build.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
Move config files to defconfigs/

Move configuration files to `defconfigs/` directory to have them all in
one place. Simplify their names and remove the preceding dot, to make
them visible to the developer / user.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
Move run scripts to run/ directory

Move run scripts to `run/` directory. This cleans up the application
root directory. Remove the `run-` prefix from their name; this combines
nicely with the `run/` directory that will prefix the run comman

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
README: Simplify clone instructions

Remove instructions to create and enter directories. `git clone`
commands create the necessary parent directories.

This simplifies the instructions in `README.md`.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
kraft.yaml: Use fs1 as root device

`fs1` is used by KraftKit as the root device. Update `kraft.yaml`
accordingly, to use `fs1` instead of `fs0`.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
redis.conf: Bind on all addresses

Update `redis.conf` configuration file to bind the Redis server to all
addresses (`0.0.0.0`), instead of just `172.44.0.2`.

This is useful when using networking setups where the address is not
beforehand and cannot be filled in the configuration file.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
Add wrapper scripts to run KraftKit

Add scripts in the `run/` directory that use `kraft run` to start
Unikraft Redis instances. Add scripts for targets in `kraft.yaml`.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
Rename fs0/ to rootfs/

`fs0` is a leftover term. Use `rootfs` as the name of the directory
storing the root filesystem. It corresponds better to its purpose.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
Use mount tag fs1 for basic builds

Use `fs1` instead of `fs0` for basic builds for consistency with
KraftKit, that defaults to `fs1` as the first mount tag.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

StefanJum commented on README.md:
````diff
@@ -55,6 +55,47 @@ OK
+redis-qemu-x86_64-initrd (qemu/x86_64)
+```
+
+Note that, if you want to build for a different architecture then the previous build, currently you are required to remove the generated files.
+For that, run the command:
+
+```console
+rm -fr .config* .unikraft/
+```
````
Is that required when using `kraft`?

razvand replied:
Yes, I get errors if I don't do that.

StefanJum commented on README.md:
````diff
@@ -173,41 +216,29 @@ Follow the steps below for the setup:
-
-     ```console
-     cd .unikraft/
+     defconfigs/  fs0/  kraft.yaml  Makefile  Makefile.uk.default  README.md  run/  workdir/
````
```suggestion
     defconfigs/  fs0/  kraft.yaml  Makefile  Makefile.uk  README.md  run/
```

StefanJum commented on README.md:
````diff
@@ -173,41 +216,29 @@ Follow the steps below for the setup:
+     git clone https://github.com/unikraft/lib-lwip workdir/libs/lwip
      ```
 
   1. Get back to the application directory:
````
```suggestion
```
This is not needed anymore

StefanJum reviewed (changes requested):
Looks good, just a few comments.

* Do we only provide basic config files? (i.e. not the `pie-paging` etc. ones)
* Would it make sense to keep the config naming convention kraft uses (`config.appname-arch...`)?

razvand commented:
> Do we only provide basic config files? (i.e. not the pie-paging etc. ones)

I will add all of them from [the `maintainer-tools` repository](https://github.com/unikraft-upb/maintainer-tools/tree/staging/workdir/copy/base/apps/redis).

> Would it make sense to keep the config naming convention kraft uses (`config.appname-arch...`)?

My thought was to simplify the names as much as possible. But if more people think we should use `config.appname-...`, then I will update accordingly. Cc: @skuenzer , @nderjung , @andreittr

StefanJum commented on README.md:
````diff
@@ -173,59 +216,41 @@ Follow the steps below for the setup:
 
      ```text
-     fs0/  kraft.yaml  Makefile  Makefile.uk.default  README.md .config.redis-qemu-x86_64-9pfs .config.redis-qemu-aarch64-9pfs [...]  run-fc-x86_64-initrd.sh*  run-qemu-aarch64-9pfs.sh*  [...]
+     defconfigs/  fs0/  kraft.yaml  Makefile  Makefile.uk  README.md [...]  run-fc-x86_64-initrd.sh*  run-qemu-aarch64-9pfs.sh*  [...]
````
```suggestion
     defconfigs/  fs0/  kraft.yaml  Makefile  Makefile.uk  README.md [...]  run/  [...]
```
These will be under the `run/` directory

StefanJum reviewed (commented):
> I will add all of them from [the maintainer-tools repository](https://github.com/unikraft-upb/maintainer-tools/tree/staging/workdir/copy/base/apps/redis).

Perfect

> My thought was to simplify the names as much as possible. But if more people think we should use config.appname-..., then I will update accordingly. 

Makes sense, I don't have a strong opinion here, just if maybe the configs can be used by kraft somehow, but that should not be the case since kraft should only care about the `kraft.yaml` file.

It looks all good on my side besides the one small comment.

razvand committed:
kraft.yaml: Update all configurations

Update all configurations in `kraft.yaml` to feature proper
configuration options. Use `arm64` instead of `aarch64` in target
names.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
defconfigs: Reduce configuration files

Reduce default configuration files (in the `deconfigs/` directory) to
their minimal configurations. Also add default configuration file for
Firecracker on AArch64.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
Introduce support for generated scripts

Add `scripts/` directory that provides support for generating build and
run scripts. Generator script is stored externally, in the `app-testing`
repository.

The `run.yaml` file defines the possible running configurations.

This deprecates the `run/` directory, so remove it.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
.gitignore: Ignore generated rootfs.cpio

Update `.gitignore` to ignore generated `rootfs.cpio` file. Previously,
the `fs0.cpio` file was generated.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

razvand committed:
README.md: Update instructions

Update instructions for building. Add instruction on using build and run
scripts. Add detailed commands on building and running.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

StefanJum reviewed (changes requested):
Either add the `setup.sh` script (and mention it in the TLDR section) or add the cloning steps back into the `TLDR` section.

razvand committed:
Introduce setup script to clone repositories

Introduce `scripts/setup.sh` script to clone repositories in the
`workdir/` directory.

Signed-off-by: Razvan Deaconescu <razvand@unikraft.io>

StefanJum reviewed (approved):
All good.
Reviewed-by: Stefan Jumarea <stefanjumarea02@gmail.com>
Approved-by: Stefan Jumarea <stefanjumarea02@gmail.com>
